### DISCUSSION TITLE
GetStatisticsOperation hung

### DISCUSSION THREAD

**Discussion by wallaceturner** [ravendb/ravendb#18217] [General]
We are making the following call to get `DatabaseStatistics` which has hung

```
DocumentStore.Maintenance.ForDatabase(DocumentStore.Database).Send(new GetStatisticsOperation());
```

(ravendb code is yellow)
![image](https://github.com/ravendb/ravendb/assets/2849980/55329679-2571-4820-8e41-78be82453a8e)

Its been hung for 18+hrs

Could you please advise if there is a possible cause and/or workaround?

**Comment by ppekrol** [comment]
Hi @wallaceturner 

Have you experienced any network issues during that period? I'm wondering if this could not cause the issue. Or is this happening each time? There is an option to force request timeout for a single call via

```
using (DocumentStore.SetRequestTimeout(value_here))
    // call here
```

or use CT in Async API.

**Comment by wallaceturner** [comment]
@ppekrol thanks for your reply
its happened once in ten years for a call that is executed once every 2 sec so we arent too concerned with it
we will investigate switching to async thx

**Comment by wallaceturner** [comment]
hi team we've had this occur again on a different client call, `session.Advanced.Stream` - see stacktrace below of the offending hang

This call *otherwise works* - i.e. its worked 99.9% of the time prior just randomly hanging once

![image](https://github.com/ravendb/ravendb/assets/2849980/1e8af81e-2349-41a1-a325-1d82d2abb46c)

code below was used before unbounded was the default in ravendb
```
        public static List<T> FetchAll<T>(this IDocumentSession session, IRavenQueryable<T> query)
        {
            var list = new List<T>();
            using (var enumerator = session.Advanced.Stream(query))
            {
                while (enumerator.MoveNext())
                {
                    list.Add(enumerator.Current.Document);
                }
            }
            return list;
        }
```

**Reply by ayende** [reply]
In this case, can you check the _tasks_ that you have running? Underneath the covers, this is done in an async task.

Do you have any non-vanilla networking setup? In general, this should either fail with a timeout or complete.

**Reply by wallaceturner** [reply]
>Do you have any non-vanilla networking setup?
No - app is running on a VM (vmware) communicating to a ravendb instance running on diff VM but same physical server

Tasks view - 
![image](https://github.com/ravendb/ravendb/assets/2849980/95c6c531-09f0-4f27-aff8-f45f032a11c3)

Should this call (`UnmanagedJsonParserHelper.ReadAsync`) have a `CancellationToken` passed down that includes a timeout? (`_cts.Token` ???) please note my knowledge of CancellationToken is rudimentary


![image](https://github.com/ravendb/ravendb/assets/2849980/cda84171-3587-48d2-9a6f-2af548ea9d2a)

**Reply by ayende** [reply]
It *should* have a token, yes. But that doesn't matter. It isn't what you are actually waiting on there, note that this is from the call to `ProcessChanges` in the `DatabaseChanges`

What this appears to be is somehow we aren't returning from the `YieldResults`, which uses the _sync_ version. 

That shouldn't matter to one  another.

What is weird is that I don't see the task for the read. There should be a `Read` operation on the `PeepingTomStream` or on the actual network stream around there.

**Reply by wallaceturner** [reply]
This has hung for us again, screenshot of the state of the Tasks is shown below and has been hung for 24+ hrs

Could you indicate a workaround or fix ?

![image](https://github.com/ravendb/ravendb/assets/2849980/9ac4e466-cc8e-4f79-8e98-a636fab32ec6)

**Reply by ayende** [reply]
Can you send us a dump file? I'm afraid that debugging via screen shots isn't really working

**Reply by wallaceturner** [reply]
https://drive.google.com/file/d/1Mp6N2nh4F4xXbYrnTvvriDs_58jdJ2TK/view?usp=sharing

**Reply by ayende** [reply]
Thanks, that was helpful, but I don't have a smoking gun yet.

Looking deeper, I have:

```
Thread   7
Current frame: 
Child-SP         RetAddr          Caller, Callee
000000107762e530 00007ffc520d1906 (MethodDesc 00007ffc51e62688 +0x56 System.Collections.Generic.GenericEqualityComparer`1[[System.__Canon, mscorlib]].Equals(System.__Canon, System.__Canon))
000000107762e7e0 00007ffbf4081ad8 (MethodDesc 00007ffbf401c8c8 +0x78 Sparrow.Json.PerCoreContainer`1[[System.__Canon, mscorlib]].TryPush(System.__Canon))
000000107762e8e0 00007ffbf431cbff (MethodDesc 00007ffbf43536f0 +0xef Raven.Client.Documents.Session.InMemoryDocumentSessionOperations.Dispose(Boolean))
000000107762e9c0 00007ffbf491042a (MethodDesc 00007ffbf3e9b760 +0x1aa FexOnline.Web.Application.Email.EmailSender.HandleMessage(FexOnline.Web.Application.Email.MailMessageArgs))
000000107762ea10 00007ffc5214f370 (MethodDesc 00007ffc51d2c930 +0xa0 System.Threading.WaitHandle.WaitAny(System.Threading.WaitHandle[], Int32, Boolean))
000000107762eb10 00007ffc5214f370 (MethodDesc 00007ffc51d2c930 +0xa0 System.Threading.WaitHandle.WaitAny(System.Threading.WaitHandle[], Int32, Boolean))
000000107762eb70 00007ffbf46271e0 (MethodDesc 00007ffbf3e9b4b0 +0x80 FexOnline.Web.Application.Framework.Threading.QueueProcessor`1[[System.__Canon, mscorlib]].Run(System.Object))
000000107762ec00 00007ffc520e00b2 (MethodDesc 00007ffc51b485b8 +0x172 System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean))
000000107762ecd0 00007ffc520dff35 (MethodDesc 00007ffc51d2c2b0 +0x15 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean))
000000107762ed00 00007ffc520dff05 (MethodDesc 00007ffc51d2c2a0 +0x55 System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object))
000000107762ed50 00007ffc52ab5700 (MethodDesc 00007ffc51d1f060 +0x60 System.Threading.ThreadHelper.ThreadStart(System.Object))
```

Which isn't shown in VS, just making sure that this isn't accidently the same session?

My issue is also that I'm seeing something *really* strange here:

```
Thread  12
Current frame: 
Child-SP         RetAddr          Caller, Callee
000000100a21d2c0 00007ffbf4063c2a (MethodDesc 00007ffbf405ef98 +0xd6a Raven.Client.Http.RequestExecutor+<SendRequestToServer>d__119`1[[System.__Canon, mscorlib]].MoveNext())
000000100a21d7f0 00007ffbf406064d (MethodDesc 00007ffbf405b9c8 +0xad System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[[Raven.Client.Http.RequestExecutor+<ExecuteAsync>d__117`1[[System.__Canon, mscorlib]], Raven.Client]](<ExecuteAsync>d__117`1<System.__Canon> ByRef))
000000100a21d830 00007ffc52152f04 (MethodDesc 00007ffc51eb5de0 +0x364 System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken))
000000100a21d910 00007ffc52152f04 (MethodDesc 00007ffc51eb5de0 +0x364 System.Threading.ManualResetEventSlim.Wait(Int32, System.Threading.CancellationToken))
000000100a21d9a0 00007ffc5214d65b (MethodDesc 00007ffc51d34508 +0xbb System.Threading.Tasks.Task.SpinThenBlockingWait(Int32, System.Threading.CancellationToken))
000000100a21da10 00007ffc52abfbf1 (MethodDesc 00007ffc51d344f8 +0x1b1 System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken))
000000100a21da30 00007ffc520d0d36 (MethodDesc 00007ffc51d0fd98 +0x16 System.AppDomain.get_Flags())
000000100a21dae0 00007ffc52ee3553 (MethodDesc 00007ffc51d1fa68 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task))
000000100a21db10 00007ffbf42e1a39 (MethodDesc 00007ffbf429d9d0 +0x89 Raven.Client.Util.AsyncHelpers.RunSync(System.Func`1<System.Threading.Tasks.Task>))
000000100a21db20 00007ffbf42e1968 (MethodDesc 00007ffbf4390b28 +0x98 Raven.Client.Http.RequestExecutor.Execute[[System.__Canon, mscorlib]](Raven.Client.Http.RavenCommand`1<System.__Canon>, Sparrow.Json.JsonOperationContext, Raven.Client.Documents.Session.SessionInfo))
000000100a21db40 00007ffbf42e138d (MethodDesc 00007ffbf43864f0 +0x4d Raven.Client.Documents.Session.Operations.QueryOperation.CreateRequest())
```

That stack trace shouldn't be possible, `ManualResetEventSlim.Wait` don't do such calls.


To rule out things, can you verify that you aren't running things concurrently with the same session?

Also, is this something that you can reproduce on a consistent or semi-consistent basis?

There is a setting that may be relevant: `AsyncHelpers.UseTaskAwaiterWhenNoSynchronizationContextIsAvailable`, but that depend on whatever you can reproduce and we can see if it helps or not.

**Reply by wallaceturner** [reply]
Thank you for taking a look

_Which isn't shown in VS, just making sure that this isn't accidently the same session?_
No that is an email sending background thread which handles its own db sessions

_To rule out things, can you verify that you aren't running things concurrently with the same session_
I will check because to be thorough this is not a 5 second answer.  The quick answer is no and the reason i say this is i know those types of things can be nasty and previously you have said 'sessions are cheap' or something along that line so if we DO need to spawn a thread to do something typically it gets a reference to `IDocumentStore` and will create its own `IDocumentSession` - thats IF we create a new thread - mostly things happen on a single thread where we dont try to re-use the session anyway (for reasons previously given)

_Also, is this something that you can reproduce on a consistent or semi-consistent basis?_
No it takes a few weeks and only manfiests when the previously mentioned 'main thread' hangs and we detect this.

**Reply by ayende** [reply]
It looks like you are able to capture that in the debugger, though? Is that happening during debug?

**Reply by wallaceturner** [reply]
ah no its the dump file created from the deployed app (via Task Manager -> Create Dump File)

**Reply by ayende** [reply]
Hm... really strange. 
Can you try `AsyncHelpers.UseTaskAwaiterWhenNoSynchronizationContextIsAvailable=false` and see if this helps?
Anything in the RavenDB logs when this happens?

As far as I can see, there should not be such a possibility.

**Reply by wallaceturner** [reply]
are there any downsides to setting `UseTaskAwaiterWhenNoSynchronizationContextIsAvailable` ? (i could not find docs on this flag)
also by ravendb logs are you referring to the studio and  `Manage Server -> Admin Logs` ?
Thanks

**Reply by ayende** [reply]
Yes, the admin logs.
And this is mostly to test out the behavior. It moves the way we are waiting for async operations for sync code to `GetAwaiter().GetResult()`, basically.

Just to be sure, you aren't messing with syncronization contexts, right?

**Reply by wallaceturner** [reply]
_Just to be sure, you aren't messing with syncronization contexts, right?_
correct, not messing with any sync contexts

**Reply by ayende** [reply]
Okay, I dug a _lot_ deeper into this.
The issue is actually related to using `StreamWithTimeout`, where we error if we didn't get anything within 2 minutes.

So that should _really_ be utterly impossible to hung.
This is probably where it is hanging:

https://github.com/ravendb/ravendb/blob/78861c6deff0d6e9fd99dc6f315c95e7bb5dbde5/src/Raven.Client/Util/StreamWithTimeout.cs#L155-L163

But you can see that in addition to the network delay, it should absolutely fail on timeout.


Looks like the query is something like:

```
from index 'Instrs/ByInstrId' where IsDeleted = $p0 order by InstrName
```

**Reply by wallaceturner** [reply]
yes that is accurate re: the query - there are ~50 documents _in total_ in that collection - the query was the first thing i checked as we could also determine which query was running due to the stack. there is nothing particulary special or heavy about that query

Is it possible this is a dotnet framework issue rather than ravendb?

**Reply by ayende** [reply]
My issue is that I can't _see_ anything that is wrong there. 
The dump you provided shows:

``
000000100a21da20 00007ffc52abfbf1 System.Threading.Tasks.Task.InternalWait(Int32, System.Threading.CancellationToken) [f:\dd\ndp\clr\src\BCL\system\threading\Tasks\Task.cs @ 3259]
000000100a21daf0 00007ffc52ee3553 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
000000100a21db20 00007ffbf42e1a39 Raven.Client.Util.AsyncHelpers.RunSync(System.Func`1<System.Threading.Tasks.Task>)
000000100a21db90 00007ffbf42e11e1 Raven.Client.Documents.Session.DocumentQuery`1[[System.__Canon, mscorlib]].ExecuteActualQuery()
```

We are waiting on task:

```
0:012> !DumpObj /d 000000105d1f48a8
Name:        System.Threading.Tasks.Task`1[[System.Threading.Tasks.VoidTaskResult, mscorlib]]
```


This is expected, we are waiting for the network call to return.
Do you perhaps define a very large timeout? The default is 15 seconds.

There is no one that is holding on to that instance, however, so no actual way to _wake_ this. 
This is really strange. 

Could you provide another dump when this happens again?

**Reply by wallaceturner** [reply]
_Do you perhaps define a very large timeout?_
we have not changed or overridden any

_Could you provide another dump when this happens again?_

will do
